frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -3956,6 +4011,145 @@ bool isTypeDefaultInitializable(Context* context, const Type* t) {
   return isTypeDefaultInitializableQuery(context, t);
 }
 
+void getCopyOrAssignableInfo(Context* context, const Type* t,
+                                    bool* fromConst, bool* fromRef,
+                                    bool checkCopyable);
+
+// checkCopyable is true for copyable, false for assignable.
+// Return: First is from const, second is from ref.
```

Comment:
The structure of this query is OK but you might also consider having a single query that computes both assignable and copyable. You'd almost certainly want to return a custom `struct` to encapsulate the results, which is just slightly harder because that struct will have to implement things like `update` / `mark` / `==` etc. Either way is OK with me but I thought I'd bring it up in case you like the alternative.

Reply:
I think that would probably be the right thing to do since the `checkCopyable` thing is a little ugly, and maybe it saves us a slight bit of redundant calculation sometimes though I'm not sure how to estimate that. I'll change it

Reply:
One concern that you haven't gotten to here yet is that it's possible to have something that supports `init=` but not `=` and it's possible to indicate an `init=` is not available by having it contain a `compilerError`. But I think these things can be future work. It's just that this query will have to respond gracefully to those situations (and in particular, it's checking for these functions, so we shouldn't get errors at this point if they are missing or contain `compilerError` calls).

Reply:
That problem would be unaddressed whether or not the query computes both assignable and copyable, right? I don't see how it's connected.

Reply:
At first I thought it was related because you might want to query if `init=` works without also checking if `=` works. But, in writing the above, I realized that either way, the query needs to keep the errors hidden, and just figure out true/false for the various possibilities.

Reply:
I see.

How would I be able to tell if the proc contains a `compilerError`? I assumed in that case it just would not resolve. And, do you think I should address it in this PR?

Reply:
Let's just note it as Future Work in this PR's PR message.

Reply:
Thinking about computing both in one query more, I'm not sure it's a good idea for compiler performance. It means we have to do two call resolutions for each of these prim calls instead of one. We only save work in the case where we check both copyability and assignability on the same type, and most of the work is probably in the resolutions that need to be done for each anyways.

In modules, the only place we check assignability for a type we also check copyability on is 4 instances in `modules/internal/ChapelSyncvar.chpl`, whereas there are ~30 instances of checking only copyability scattered through our standard collection types.

I'm most of the way through creating a custom struct to represent copyability and assignability together, and could still use it to just represent one or the other at a time, but at that point I think the advantage over the pair of bools isn't worth the additional weight.

Reply:
The thing is, we have in the compiler some errors that appear if you have say an `init=` but not an `=`. So resolving these all might not be extra work once we have that error checking. (Open question if the query you are working on would be an appropriate place to do this error checking, though).

Reply:
I see. I think this query is not an appropriate place to check for it, but perhaps both this query and the place that does the checking could call a common query and that work wouldn't be redundant. However, I think the error is only if we had a user-defined `init=` and no user-defined `=`, and for the purposes of this query we don't care if we find something user-defined or compiler-generated, so it would take more refactoring in the future that we might not find practical to make them overlap.

For this PR I think I should add and use the custom struct just to represent either copyability or assignability (with a flag for which), and keep the query just calculating one or the other at once. That would make it pretty easy to extend to calculating them both together in the future, without any performance impact for now. Does that sound good?

Reply:
> Does that sound good?

Works for me!